javaslang/src/main/java/javaslang/collection/Array.java

```diff
@@ -807,23 +801,25 @@ public int length() {
     public Array<Array<T>> permutations() {
         if (isEmpty()) {
             return empty();
+        } else if (length() == 1) {
+            return of(this);
         } else {
-            final Array<T> tail = tail();
-            if (tail.isEmpty()) {
-                return of(this);
-            } else {
-                final Array<Array<T>> zero = empty();
-                return distinct().foldLeft(zero, (xs, x) -> {
-                    final Function<Array<T>, Array<T>> prepend = l -> l.prepend(x);
-                    return xs.appendAll(remove(x).permutations().map(prepend));
-                });
+            Array<Array<T>> results = empty();
+            for (T t : distinct()) {
+                for (Array<T> ts : remove(t).permutations()) {
+                    results = results.append(of(t).appendAll(ts));
+                }
             }
+            return results;
         }
     }
 
     @Override
     public Array<T> prepend(T element) {
-        return insert(0, element);
+        final Object[] copy = new Object[length() + 1];
```

Comment:
Maybe we should move this code to the `insert` method and add an `if-else` for the case that the index is `0`. This could also boost the insert operation a little. What do you think?